plugins/woocommerce/tests/php/src/Internal/ProductFilters/AbstractProductFiltersTest.php

```diff
@@ -0,0 +1,329 @@
+				'stock_status' => 'instock',
+				'variations'   => array(
+					array(
+						'attributes' => array(
+							'pa_color' => 'blue',
+						),
+						'props'      => array(
+							'regular_price' => 60,
+							'stock_status'  => 'instock',
+						),
+					),
+					array(
+						'attributes' => array(
+							'pa_color' => 'green',
+						),
+						'props'      => array(
+							'regular_price' => 60,
+							'stock_status'  => 'instock',
+						),
+					),
+				),
+			),
+		);
+
+		$this->products = array_map(
+			array( $this, 'create_test_product' ),
+			$this->products_data
+		);
+	}
+
+	/**
+	 * Runs after each test.
+	 */
+	public function tearDown(): void {
+		parent::tearDown();
+
+		$this->remove_all_attributes();
+		$this->remove_all_products();
+		$this->empty_lookup_tables();
+		wp_cache_flush();
+	}
+
+	/**
+	 * Truncate the lookup table.
+	 */
+	private function empty_lookup_tables() {
+		global $wpdb;
+		$wpdb->query( "TRUNCATE TABLE {$wpdb->prefix}wc_product_meta_lookup" );
+		$wpdb->query( "TRUNCATE TABLE {$wpdb->prefix}wc_product_attributes_lookup" );
+	}
+
+	/**
+	 * Remove all attributes and associated terms.
+	 */
+	private function remove_all_attributes() {
+		$attribute_ids_by_name = wc_get_attribute_taxonomy_ids();
+		foreach ( $attribute_ids_by_name as $attribute_name => $attribute_id ) {
+			$attribute_name  = wc_sanitize_taxonomy_name( $attribute_name );
+			$taxonomy_name   = wc_attribute_taxonomy_name( $attribute_name );
+			$attribute_terms = get_terms( array( 'taxonomy' => $taxonomy_name ) );
+			if ( ! is_wp_error( $attribute_terms ) ) {
+				foreach ( $attribute_terms as $term ) {
+					wp_delete_term( $term->term_id, $taxonomy_name );
+				}
+			}
+			unregister_taxonomy( $taxonomy_name );
+
+			wc_delete_attribute( $attribute_id );
+		}
+	}
+
+	/**
+	 * Remove all products.
+	 */
+	private function remove_all_products() {
+		$product_ids = wc_get_products( array( 'return' => 'ids' ) );
+		foreach ( $product_ids as $product_id ) {
+			$product     = wc_get_product( $product_id );
+			$is_variable = $product->is_type( 'variable' );
+
+			foreach ( $product->get_children() as $child_id ) {
+				$child = wc_get_product( $child_id );
+				if ( empty( $child ) ) {
+					continue;
+				}
+
+				if ( $is_variable ) {
+					$child->delete( true );
+				} else {
+					$child->set_parent_id( 0 );
+					$child->save();
+				}
+			}
+
+			$product->delete( true );
+		}
+	}
+
+	/**
+	 * Get data from results of wc_get_product(), default to return the product name.
+	 *
+	 * @param \WC_Product[] $products Array of products.
+	 * @param function      $callback The callback that passed to array map.
+	 */
+	protected function get_data_from_products_array( $products, $callback = null ) {
+		if ( ! $callback ) {
+			$callback = function ( $product ) {
+				return $product->get_name();
+			};
+		}
+
+		return array_map(
+			$callback,
+			$products
+		);
+	}
+
+	/**
+	 * Build and create attribute from variations data.
```

Comment:
I guess this should be a plural?

```suggestion
	 * Build and create attributes from variations data.
```